### DISCUSSION TITLE
Memory Allocation Query

### DISCUSSION THREAD

**Discussion by AubsUK** [egzumer/uv-k5-firmware-custom#468] [General]
Looking at `settings.c`, and part of the `SETTINGS_FactoryReset` function:
```
(
	!(i >= 0x0D60 && i < 0x0E28) &&     // MR Channel Attributes
	!(i >= 0x0F18 && i < 0x0F30) &&     // Scan List
	!(i >= 0x0F50 && i < 0x1C00) &&     // MR Channel Names
	!(i >= 0x0E40 && i < 0x0E70) &&     // FM Channels
	!(i >= 0x0E88 && i < 0x0E90)        // FM settings
)

```

That matches with one section in the `SETTINGS_InitEEPROM` function also in `settings.c`:
```
// 0D60..0E27
EEPROM_ReadBuffer(0x0D60, gMR_ChannelAttributes, sizeof(gMR_ChannelAttributes));
```

Except the following declaration from `misc.h` indicates each channel att contains 2 bytes:

```
typedef union {
    struct {
        uint8_t
            band : 4,
            compander : 2,
            scanlist2 : 1,
            scanlist1 : 1;
    };
    uint8_t __val;
} ChannelAttributes_t;
```
and the declaration also in `misc.h` indicates there's 207 of them:
```
extern ChannelAttributes_t   gMR_ChannelAttributes[207];
```

Surely that would mean it needs 207 objects * 2 bytes each = **414 bytes**, but 0D60..0E27 represents 3424..3623 = **200 bytes** - One for each channel, not the required two.

But then also in the `SETTINGS_InitEEPROM` function in `settings.c`:
```
	for(uint16_t i = 0; i < sizeof(gMR_ChannelAttributes); i++) {
```
Doesn't that mean it reads 16-bit (2 bytes) for each channel?

I'm clearly missing something, could someone help clarify this for me please?

**Comment by egzumer** [comment]
Attributes are one byte long.
207 is 200 channels + 7 bands in frequency mode

[UV K5 EEPROM.ods](https://github.com/egzumer/uv-k5-firmware-custom/files/14234234/UV.K5.EEPROM.ods)

**Reply by AubsUK** [reply]
The memory space map is really useful thank you.

I realise I was wrong because the `struct { band / compander / scanlist2 / scanlist1 }` and `__val` occupy the same 8-bit memory location.
